File: Day237_2025/Q237_Leetcode.java
import java.util.*;

class Q237_Leetcode {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        Map<Integer, List<Integer>> map = new TreeMap<>();
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                map.computeIfAbsent(i + j, _ -> new ArrayList<>()).add(mat[i][j]);
            }
        }
        
        List<Integer> resList = new ArrayList<>();
        for (Map.Entry<Integer, List<Integer>> entry : map.entrySet()) {
            List<Integer> diag = entry.getValue();
            if (entry.getKey() % 2 == 0) Collections.reverse(diag);
            resList.addAll(diag);
        }
        
        int[] res = new int[resList.size()];
        for (int i = 0; i < res.length; i++) res[i] = resList.get(i);
        return res;
    }
}

File: Day237_2025/Q237_Leetcode.py
from typing import List
from collections import defaultdict

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        diag_map = defaultdict(list)
        m, n = len(mat), len(mat[0])
        
        for i in range(m):
            for j in range(n):
                diag_map[i + j].append(mat[i][j])
        
        res = []
        for k in range(m + n - 1):
            diag = diag_map[k]
            if k % 2 == 0:
                diag.reverse()
            res.extend(diag)
        
        return res
